src/renderer/generic.zig

```diff
@@ -0,0 +1,2873 @@
+        /// but we keep this around so that we don't reallocate. Each set of
+        /// cells goes into a separate shader.
+        cells: cellpkg.Contents,
+
+        /// The last viewport that we based our rebuild off of. If this changes,
+        /// then we do a full rebuild of the cells. The pointer values in this pin
+        /// are NOT SAFE to read because they may be modified, freed, etc from the
+        /// termio thread. We treat the pointers as integers for comparison only.
+        cells_viewport: ?terminal.Pin = null,
+
+        /// Set to true after rebuildCells is called. This can be used
+        /// to determine if any possible changes have been made to the
+        /// cells for the draw call.
+        cells_rebuilt: bool = false,
+
+        /// The current GPU uniform values.
+        uniforms: shaderpkg.Uniforms,
+
+        /// The font structures.
+        font_grid: *font.SharedGrid,
+        font_shaper: font.Shaper,
+        font_shaper_cache: font.ShaperCache,
+
+        /// The images that we may render.
+        images: ImageMap = .{},
+        image_placements: ImagePlacementList = .{},
+        image_bg_end: u32 = 0,
+        image_text_end: u32 = 0,
+        image_virtual: bool = false,
+
+        /// Graphics API state.
+        api: GraphicsAPI,
+
+        /// The CVDisplayLink used to drive the rendering loop in
+        /// sync with the display. This is void on platforms that
+        /// don't support a display link.
+        display_link: ?DisplayLink = null,
+
+        /// Health of the most recently completed frame.
+        health: std.atomic.Value(Health) = .{ .raw = .healthy },
+
+        /// Our swap chain (multiple buffering)
+        swap_chain: SwapChain,
+
+        /// This value is used to force-update swap chain targets in the
+        /// event of a config change that requires it (such as blending mode).
+        target_config_modified: usize = 0,
+
+        /// If something happened that requires us to reinitialize our shaders,
+        /// this is set to true so that we can do that whenever possible.
+        reinitialize_shaders: bool = false,
+
+        /// Whether or not we have custom shaders.
+        has_custom_shaders: bool = false,
+
+        /// Our shader pipelines.
+        shaders: Shaders,
+
+        /// Swap chain which maintains multiple copies of the state needed to
+        /// render a frame, so that we can start building the next frame while
+        /// the previous frame is still being processed on the GPU.
+        const SwapChain = struct {
+            // The count of buffers we use for double/triple buffering.
+            // If this is one then we don't do any double+ buffering at all.
+            // This is comptime because there isn't a good reason to change
+            // this at runtime and there is a lot of complexity to support it.
+            const buf_count = count: {
+                if (@hasDecl(GraphicsAPI, "swap_chain_count")) {
```

Comment:
I'm doing my best to move away from conditional `hasDecl` where possible because it makes the "comptime interface" more mysterious and implicit. It does make sense in some cases where the behavior is super rare, but I'd lean towards avoiding by default.

In this case, I think we should make `swap_chain_count` a required decl on GraphicsAPI to ensure our renderer authors think about it.